compiler/rustc_const_eval/src/interpret/discriminant.rs

```diff
@@ -166,31 +166,59 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                         let variant_index_relative =
                             variant_index_relative_val.to_scalar().to_bits(tag_val.layout.size)?;
                         // Check if this is in the range that indicates an actual discriminant.
-                        if variant_index_relative <= u128::from(variants_end - variants_start) {
-                            let variant_index_relative = u32::try_from(variant_index_relative)
-                                .expect("we checked that this fits into a u32");
-                            // Then computing the absolute variant idx should not overflow any more.
-                            let variant_index = VariantIdx::from_u32(
-                                variants_start
-                                    .checked_add(variant_index_relative)
-                                    .expect("overflow computing absolute variant idx"),
-                            );
-                            let variants =
-                                ty.ty_adt_def().expect("tagged layout for non adt").variants();
-                            assert!(variant_index < variants.next_index());
-                            variant_index
+                        if niche_variants.contains(&untagged_variant) {
+                            if variant_index_relative < u128::from(discr_len) {
+                                let adj_untagged_idx = untagged_variant.as_u32() - variants_start;
+                                let variant_index_relative = u32::try_from(variant_index_relative)
+                                    .expect("we checked that this fits into a u32");
+                                let variant_index_to_modulo = variant_index_relative
+                                    .checked_add(1)
+                                    .expect("overflow computing absolute variant idx")
+                                    .checked_add(adj_untagged_idx)
+                                    .expect("overflow computing absolute variant idx");
+                                let variant_index = VariantIdx::from_u32(
+                                    variants_start
+                                        .checked_add(variant_index_to_modulo % discr_len)
+                                        .expect("overflow computing absolute variant idx"),
+                                );
+                                let variants =
+                                    ty.ty_adt_def().expect("tagged layout for non adt").variants();
+                                assert!(variant_index < variants.next_index());
+                                variant_index
+                            } else {
+                                untagged_variant
+                            }
                         } else {
-                            untagged_variant
+                            if variant_index_relative < u128::from(discr_len) {
```

Comment:
It looks like there is some duplication of logic here? The `if variant_index_relative < u128::from(discr_len) { ... } else { untagged_variant }` occurs twice. Can that be cleaned up?

Reply:
They have different conditions. I deleted a subtraction expression by mistake. 
This reinforces the need for more thorough test coverage.